amethyst_input/src/input_handler.rs

```diff
@@ -628,3 +628,514 @@ where
         event_handler.iter_write(events);
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use std::fmt::Debug;
+
+    use super::*;
+    use winit::{
+        DeviceId, ElementState, Event, KeyboardInput, ModifiersState, WindowEvent, WindowId,
+    };
+
+    /// Compares two sets for equality, but not the order
+    fn sets_equal<T>(a: &[T], b: &[T])
+    where
+        T: PartialEq<T> + Debug,
+    {
+        let mut ret = a.len() == b.len();
+
+        if ret {
+            let mut b = b.iter().collect::<Vec<_>>();
+            for a in a.iter() {
+                if let Some(i) = b.iter().position(|b| a == *b) {
+                    b.swap_remove(i);
+                } else {
+                    ret = false;
+                    break;
+                }
+            }
+            assert_eq!(b.len(), 0);
```

Comment:
Ah, this would prevent us from getting the nice error message below when there are unequal types &ndash; we'd get

```
left: `2` right: `0`
```

or similar. If you remove this line then we'd get what we should.

Reply:
Agreed. Something like this is probably better:
```rust
assert_eq!(a.iter().collect::<HashSet<_>>(), b.iter().collect::<HashSet<_>>());
```